rfcs/rfc-0004-text-formatting-extensions.md

```diff
@@ -0,0 +1,371 @@
+
+To include math formulas in your messages, you can continue using this math syntax:
+
+```markdown
+$$
+\sqrt{3x-1}+(1+x)^2
+$$
+```
+
+<!-- Explain the proposal as if it's already in Revolt and you were teaching it to new users.
+- Introduce new concepts
+- Explain the feature with examples
+- What this fixes or adds and what users should think of the feature
+- Discuss how this impacts using Revolt, how it makes it harder or easier to use
+
+For internal oriented RFCs such as internal code changes, this should largely talk about how contributors should think about the change and give examples on the impacts. -->
+
+# Reference-level explanation
+
+The text formatting extension is a simple extension to RSM that allows for the formatting of text in a more expressive way. It is designed to be simple to implement and amend, and is designed to be consistent with the rest of the client's UI as well as allow for adjustments between platforms.
+
+The text formatting extension is implemented as a double-brace syntax, with the first brace (STYLE BRACE), denoted by curly braces, wherein `{` is the left brace and `}` is the right brace, containing the formatting options. Immediately following without any whitespace is the second brace (CONTENT BRACE), denoted by parentheses, wherein `(` is the left brace and `)` is the right brace, containing the text to be formatted. The text to be formatted is allowed to contain any characters and is not subject to special restrictions.
+
+```markdown
+{STYLE BRACE}(CONTENT BRACE)
+```
+
+The STYLE BRACE contains the formatting options, which are separated by semicolons. The formatting options are key-value pairs, with the key and value separated by a colon. The key is the name of the formatting option, and the value is the value of the formatting option. The key is **case-insensitive**, and the value is **case-sensitive**. The key and value are separated by a colon, and the key-value pairs are separated by semicolons. The key is not allowed to contain any characters other than letters and the spe
+, and the value is allowed to contain any characters with the exception of semicolons and closing braces. The key is not allowed to be empty, and the value is not allowed to be empty. The key is not allowed to be repeated. Keys that do not exist are ignored silently in formatting, but clients may choose to display a warning to the user sending the message, or to the user receiving the message, or both. Trailing semicolons are not allowed in the STYLE BRACE. The STYLE BRACE is not allowed to be empty.
+
+We can deduce the following [grammar](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form) for the STYLE BRACE:
+
+```bnf
+STYLE_BRACE ::= "{" STYLE_PAIR "}"
+STYLE_PAIR ::= STYLE_PAIR ";" STYLE_PAIR | STYLE_PAIR
+STYLE_PAIR ::= KEY ":" VALUE
+KEY ::= [a-zA-Z]+
+VALUE ::= [^;]+
+```
```

Comment:
This doesn't seem to account for the whitespaces here?
```
{color: red; font: monospace; background: yellow}(red)
            ^                ^                        
```

My current regex looks like this:
```js
/\{([a-zA-Z]+):([^;]+?)(?:;([a-zA-Z]+):([^;]+?))*\}\(/g
```

Based off the spec, but this doesn't match string provided.